astropy/coordinates/sky_coordinate.py

```diff
@@ -65,19 +66,32 @@ def _repr_data(self):
         return repr_data
 
     def _represent_as_dict(self):
-        obj = self._parent
-        attrs = (list(obj.representation_component_names) +
-                 list(frame_transform_graph.frame_attributes.keys()))
-
-        # Don't output distance if it is all unitless 1.0
-        if 'distance' in attrs and np.all(obj.distance == 1.0):
-            attrs.remove('distance')
+        sc = self._parent
+        attrs = list(sc.representation_component_names)
+
+        # Don't output distance unless it's actually distance.
+        if issubclass(sc.data.__class__, UnitSphericalRepresentation):
+            attrs = attrs[:-1]
+
+        diff = sc.data.differentials.get('s')
+        if diff is not None:
+            diff_attrs = list(sc.get_representation_component_names('s'))
+            # Don't output proper motions if they haven't been specified.
+            if issubclass(diff.base_representation, RadialRepresentation):
+                diff_attrs = diff_attrs[2:]
+            # Don't output radial velocity unless it's actually velocity.
+            elif issubclass(diff.base_representation,
+                            UnitSphericalRepresentation):
+                diff_attrs = diff_attrs[:-1]
+            attrs.extend(diff_attrs)
+
+        attrs.extend(list(frame_transform_graph.frame_attributes.keys()))
```

Comment:
In principle no need to to cast to `list` here - the `.extend` method takes any iterable.